### PR TITLE
Add support for KeyValuePairExtractionScore evaluator

### PR DESCRIPTION
**Pull request by juliagongms**
Adds support for evaluation of the key-value pair extraction task with per-key evaluation for simple and complex object types.

### CODE REVIEW

File: visionmetrics/key_value_pair/key_value_pair_eval_base.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    It accepts a dictionary mapping the keys to corresponding metric information, and returns the corresponding key-wise set of metrics.
+    Each key can have a different metric for evaluation. The metrics supported are specified in SupportedKeyWiseMetric.
+
+    Args:
+        key_metric_map: dictionary from keys (extracted field names) to a dictionary with three required fields:
+        1. metric_name: string of the metric name as defined in visionmetrics (e.g., 'classification.MulticlassAccuracy'), which should be among the ones specified in SupportedKeyWiseMetric.
+        2. metric_args: dictionary of args to pass in as keyword arguments to the initialization function of the metric.
+        3. preprocessor: function object that can be called with (prediction, target) values for a single instance to preprocess them into the desired format for the corresponding metric.
+        4. key_trace: list of strings of key names that traces the path to the current key in the key-value pair prediction/target object (not in the schema).
```
Comment (nisyad-ms) [inline visionmetrics/key_value_pair/key_value_pair_eval_base.py:35]: The meaning of `key_trace` is not immediately obvious. A simple example might help.

File: visionmetrics/key_value_pair/key_value_pair_eval_base.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            except Exception as e:
+                raise ValueError(f"Encountered error '{repr(e)}' when instantiating metric '{metric_name}' for key '{key}' with arguments '{metric_args}'.")
+
+    def update(self, predictions, targets):
+        """
+        Updates metrics for each key using all samples in predictions and targets.
+        Both predictions and targets should be dictionaries of the form {'<key>': <value>}, where <value> is in the format expected for the respective metric for that key.
+        Each sample in predictions and targets must have the same keys (though they do not have to have all the keys in the dataset).
+        """
+        for key in self.key_evaluator_map:
```
Comment (cy-bai) [inline visionmetrics/key_value_pair/key_value_pair_eval_base.py:110]: nit: `for key, key_metric in self.key_evaluator_map.items():`

File: visionmetrics/key_value_pair/key_value_pair_eval_base.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            metric = self.key_evaluator_map[key]
+            metric_name = self.key_metric_map[key]["metric_name"]
+            key_predictions = []
```
Comment (cy-bai) [inline visionmetrics/key_value_pair/key_value_pair_eval_base.py:113]: `key_trace = self.key_metric_map[key]["key_trace"]`
Then reuse this variable in below for loop

File: visionmetrics/key_value_pair/key_value_pair_eval.py [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+            key_trace: list of strings of key names that traces the path to the current key in the key-value pair prediction/target object (not in the schema).
+        """
+        # Use text as the default metric for all keys
+        self._assign_key_metric_map_values(key=key,
+                                           metric_name=SupportedKeyWiseMetric.Caption_AzureOpenAITextModelCategoricalScore,
+                                           metric_args={"endpoint": self.endpoint, "deployment_name": self.deployment_name})
+        if key_schema["type"] == JSONSchemaKeyType.String:
+            if "enum" in key_schema:
+                class_map = self._get_enum_class_map(key_schema["enum"])
+                self._assign_key_metric_map_values(key=key,
```
Comment (nisyad-ms) [inline visionmetrics/key_value_pair/key_value_pair_eval.py:149]: nit: instead of repeated calls to `self._assign_key_metric_map_values`, may initiate empty mapping dict at the top and fill it as per every IF/ELSE condition and call `self._assign_key_metric_map_values(**mapping)` at the end. 
If mapping is empty after all IF/ELSE, then initialize it with the default metric

File: visionmetrics/key_value_pair/key_value_pair_eval_base.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+            except KeyError:
+                raise ValueError(f"Each value in key_metric_map must be a dictionary with a 'metric_name' key for a supported metric in visionmetrics of the form <task_name>.<metric_name>"
+                                 f" and a 'metric_args' key for a dictionary of arguments to pass to the initialization function of that metric. Key '{key}' does not satisfy this rule.")
+            if metric_name not in [m.value for m in SupportedKeyWiseMetric]:
+                raise ValueError(f"Metric '{key_metric_map[key]['metric_name']}' is not supported. "
+                                 f"Each key's metric must be among the supported metrics: {', '.join([m.value for m in SupportedKeyWiseMetric])}.")
+
+            # Instantiate metric
+            try:
+                self.key_evaluator_map[key] = eval(f"{metric_name}(**metric_args)")
```
Comment (nisyad-ms) [inline visionmetrics/key_value_pair/key_value_pair_eval_base.py:100]: I think you meant metric_name.value? (metric_name is an Enum and not callable, unittest failed)

File: visionmetrics/key_value_pair/key_value_pair_eval_base.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    """
+    Evaluator for evaluating key-value pair datasets, where each key in the labels represents an extracted field with a consistent type (string, number, etc.).
+    It accepts a dictionary mapping the keys to corresponding metric information, and returns the corresponding key-wise set of metrics.
+    Each key can have a different metric for evaluation. The metrics supported are specified in SupportedKeyWiseMetric.
+
+    Args:
+        key_metric_map: dictionary from keys (extracted field names) to a dictionary with four required fields:
+        1. metric_name: string of the metric name as defined in visionmetrics (e.g., 'classification.MulticlassAccuracy'), which should be among the ones specified in SupportedKeyWiseMetric.
+        2. metric_args: dictionary of args to pass in as keyword arguments to the initialization function of the metric.
+        3. preprocessor: function object that can be called with (prediction, target) values for a single instance to preprocess them into the desired format for the corresponding metric.
```
Comment (cy-bai) [inline visionmetrics/key_value_pair/key_value_pair_eval_base.py:34]: do we expect preprocessor to required coupled gt and pred? Otherwise, can we separate to gt_preprocessor and pred_preprocessor?  In some cases they are same or one can be adapted from another

File: visionmetrics/key_value_pair/key_value_pair_eval_base.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        """
+        for key in self.key_evaluator_map:
+            metric = self.key_evaluator_map[key]
+            metric_name = self.key_metric_map[key]["metric_name"]
+            key_predictions = []
+            key_targets = []
+            for prediction, target in zip(predictions, targets):
+                # Use the key trace to traverse the prediction and target objects to get the values for these keys
+                try:
+                    key_prediction = reduce(operator.getitem, self.key_metric_map[key]["key_trace"], prediction)
```
Comment (nisyad-ms) [inline visionmetrics/key_value_pair/key_value_pair_eval_base.py:118]: nit: maybe extract the repeated reduce operation to minimize duplicate code

File: tests/test_key_value_pair.py [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+    # In each unit test, we don"t unit test the preprocessor name explicitly because we will test it implicitly via the "update" function call tests.
+    def test_key_value_pair_extraction_evaluator_simple_schema(self):
+        evaluator = KeyValuePairExtractionScore(key_value_pair_schema=self.simple_schema,
+                                                endpoint=self.ENDPOINT,
+                                                deployment_name=self.DEPLOYMENT)
+        with unittest.mock.patch.object(OpenAITextChatModel, "forward", return_value=["1.0"]):
+            for key in self.simple_key_metric_map:
+                for field in self.simple_key_metric_map[key]:
+                    self.assertEqual(evaluator.key_metric_map[key][field], self.simple_key_metric_map[key][field])
+            evaluator.update(predictions=[{
```
Comment (nisyad-ms) [inline tests/test_key_value_pair.py:274]: Please add a test case where the prediction and targets keys do not match.

File: visionmetrics/key_value_pair/key_value_pair_eval_base.py (hunk 7 above)
Comment (juliagongms) [reply visionmetrics/key_value_pair/key_value_pair_eval_base.py:118]: Hm just to make sure I understand, the reduce operation is on two different objects (prediction and target). Are you suggesting to create a partial lambda function that just takes prediction/target?

File: visionmetrics/key_value_pair/key_value_pair_eval_base.py (hunk 3 above)
Comment (juliagongms) [reply visionmetrics/key_value_pair/key_value_pair_eval_base.py:113]: Good catch, thanks!

File: visionmetrics/key_value_pair/key_value_pair_eval.py [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+                                                   metric_name=SupportedKeyWiseMetric.Classification_MulticlassF1,
+                                                   metric_args={"num_classes": len(class_map), "average": "micro"},
+                                                   class_map=class_map)
+        elif key_schema["type"] == JSONSchemaKeyType.Number or key_schema["type"] == JSONSchemaKeyType.Integer:
+            self._assign_key_metric_map_values(key=key,
```
[the remark this answers is not in the document]
Comment (juliagongms) [reply visionmetrics/key_value_pair/key_value_pair_eval.py:154]: I believe the `elif` branch only runs if the previous statements failed (similar to `else if` in OO languages like Java). Is this what you mean? If so, I think this code already accomplishes this, but let me know if I'm misunderstanding.

File: visionmetrics/key_value_pair/key_value_pair_eval_base.py (hunk 5 above)
Comment (juliagongms) [reply visionmetrics/key_value_pair/key_value_pair_eval_base.py:115]: Hm, odd. The value of `metric_name` is the enum value (not the Enum object), so the string should be something like "classification.MulticlassAccuracy". The unit test in the CI also appears to be working (works for me locally as well).

File: visionmetrics/key_value_pair/key_value_pair_eval.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (cy-bai) [reply visionmetrics/key_value_pair/key_value_pair_eval.py:155]: If I understand correctly, for number/integer, the metric is either MAE (non-enum) or F1 (enum),  so could we do:
`if "enum" in key_shema:
    # assign F1
else:
   # assign MAE
`
current implementation has to assign metric twice for enum type.

File: visionmetrics/key_value_pair/key_value_pair_eval_base.py (hunk 7 above)
Comment (nisyad-ms) [reply visionmetrics/key_value_pair/key_value_pair_eval_base.py]: Yes, but on second thoughts it may not be worth it. Just makes it harder to debug. I will mark this as resolved.

File: visionmetrics/regression/accuracy.py [hunk 10]
```diff
@@ -8,19 +8,17 @@ class MeanAbsoluteErrorF1Score(Metric):
     Args:
         threshold: float indicating the exclusive threshold below which the mean absolute error is considered a true positive.
```
Comment (nisyad-ms) [inline visionmetrics/regression/accuracy.py:9]: super irritating nit: change docstring arg to error_threshold.

File: visionmetrics/key_value_pair/key_value_pair_eval_base.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+                    continue
+
+                # Construct expected evaluation metric update format for the current key
+                try:
+                    preprocessor = self.key_metric_map[key]["preprocessor"]
+                    key_prediction_formatted, key_target_formatted = preprocessor(key_prediction, key_target)
+                    key_predictions.append(key_prediction_formatted)
+                    key_targets.append(key_target_formatted)
+                except ValueError as e:
+                    logger.debug(f"Encountered error {repr(e)} when preprocessing prediction '{key_prediction}' and target '{key_target}' for key '{key}' to the metric's expected format.")
```
[the remark this answers is not in the document]
Comment (juliagongms) [reply visionmetrics/key_value_pair/key_value_pair_eval_base.py:152]: Agreed! I originally implemented it this way in case we wanted to run the metric even in cases where the output values might be malformed, but on second thought, since we expect the outputs to be well-formed 100% of the time, it would make more sense just to throw an error.

File: visionmetrics/regression/accuracy.py (hunk 10 above)
Comment (juliagongms) [reply visionmetrics/regression/accuracy.py:9]: Ahh, great catch, didn't update this after renaming it in the code. Thanks!

File: visionmetrics/key_value_pair/key_value_pair_eval.py [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+        Recursive function that populates the object-level key_metric_map dictionary. For a given key and key_schema in JSON Schema format,
+        the function populates the metric_name, metric_args, and preprocessor values in the dictionary.
+        Note: The recursive behavior currently only occurs for JSONSchemaKeyType.Object key types, not arrays.
+        Args:
+            key: string of the key name, which should be identical to the key that will be evaluated in the prediction and target dictionaries.
+            key_schema: dictionary in JSON schema format specifying the expected schema for the prediction and target dictionaries to be used in evaluation.
+            key_trace: list of strings of key names that traces the path to the current key in the key-value pair prediction/target object (not in the schema).
+        """
+        # Use text as the default metric for all keys
+        self._assign_key_metric_map_values(key=key,
```
Comment (nisyad-ms) [inline visionmetrics/key_value_pair/key_value_pair_eval.py:144]: How about we assign the default values only when all IF/ELSE conditions are not satisfied and log this information for debugging that "XYZ is missing/not found...using default values...". 
Currently, assigning a default value and then overwriting it later appears brittle. Could even be more error prone in future. 
+ @cy-bai for his thoughts.

Comment (juliagongms) [reply visionmetrics/key_value_pair/key_value_pair_eval.py:144]: Good idea, thanks!

File: visionmetrics/key_value_pair/key_value_pair_eval_base.py (hunk 6 above)
Comment (juliagongms) [reply visionmetrics/key_value_pair/key_value_pair_eval_base.py:49]: Good idea to decouple them, thanks!

**Review by nisyad-ms — APPROVED** [review]

File: visionmetrics/regression/accuracy.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+    def update(self, predictions, targets):
+        if not torch.is_tensor(predictions):
+            raise ValueError("'predictions' must be a float or integer torch.tensor.")
+        if not torch.is_tensor(targets):
+            raise ValueError("'targets' must be a float or integer torch.tensor.")
+        if predictions.shape != targets.shape:
+            raise ValueError(f"'predictions' and 'targets' must have the same shape; got predictions of shape {list(predictions.shape)} and targets of shape {list(targets.shape)}.")
+        n_obs = targets.numel()
+        tp = torch.sum(torch.where(torch.abs(predictions - targets) < self.error_threshold, 1, 0))
```
Comment (nisyad-ms) [inline visionmetrics/regression/accuracy.py:29]: Would it make more sense to set default value of error_threshold = 0 instead of 1.0. As a new user I may end up using the default metric expecting the usual MAE error. This way one has to explicitly set an error threshold. 

In line 29, we can then use a strict <= comparison. 
Also, maybe add a condition check to ensure error_threshold > 0?

File: visionmetrics/key_value_pair/key_value_pair_eval.py [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+
+
+class KeyValuePairExtractionScore(KeyValuePairEvaluatorBase):
+    """
+    Evaluator for a key-value pair dataset. The default evaluator (if the key does not have values of a particular form) is the caption.AzureOpenAITextModelCategoricalScore due to its flexibility.
+    Note: Currently, detection metrics are class-agnostic; bounding boxes are evaluated separately from classes. TODO: Consider supporting multiclass detection metrics, which may complicate logic.
+    Args:
+        key_value_pair_schema: dictionary in JSON Schema format indicating each key and expected value type for each extracted field.
+        Examples (corresponding to the examples in key_value_pair_eval_base.py):
+        defect_detection_schema = {
```
[the remark this answers is not in the document]
Comment (juliagongms) [reply visionmetrics/key_value_pair/key_value_pair_eval.py:34]: We now support this after updating to the latest schema. Thanks!

File: visionmetrics/key_value_pair/key_value_pair_eval.py [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+
+# Constants for predictions and targets
+VALUE_SUBKEY = "value"
+GROUNDINGS_SUBKEY = "groundings"
+DESCRIPTION_SUBKEY = "description"
+
+
+class KeyValuePairExtractionScore(KeyValuePairEvaluatorBase):
+    """
+    Evaluator for a key-value pair dataset. The default evaluator (if the key does not have values of a particular form) is the caption.AzureOpenAITextModelCategoricalScore due to its flexibility.
```
Comment (pj-ms) [inline visionmetrics/key_value_pair/key_value_pair_eval.py:30]: Good to forget all your existing knowledge about vision-datasets and see if this documentation still makes sense to determine whether the documentation is helpful or not

File: README.md [hunk 16]
```diff
@@ -50,9 +49,12 @@ The above metrics are available for Binary, Multiclass, and Multilabel classific
   - `RetrievalMAP`: computes [Mean Average Precision@k](https://stackoverflow.com/questions/54966320/mapk-computation), an information retrieval metric.
   - `RetrievalPrecisionRecallCurveNPoints`: computes a Precision-Recall Curve, interpolated at k points and averaged over all samples. 
 
-### Grounding
+### Grounding:
   - `Recall`: computes Recall@k, which is the percentage of correct grounding in top-k among all relevant items.
 
+### Key-Value Pair Extraction:
+  - `KeyValuePairExtractionScore`: computes key-wise metrics for each key in a JSON Schema-formatted key-value pair schema defining a set of keys to be extracted from an image, where each key-wise metric is inferred according to the key-value pair schema.
```
Comment (pj-ms) [inline README.md:56]: probably need some more documentation here

File: visionmetrics/key_value_pair/key_value_pair_eval.py [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+    Args:
+        key_value_pair_schema: dictionary in JSON Schema format indicating each key and expected value type for each extracted field.
+        Examples (corresponding to the examples in key_value_pair_eval_base.py):
+        defect_detection_schema = {
+            "defects": {
+                "type": "array",
+                "description": "The defect types present in the image.",
+                "items": {
+                    "type": "string",
+                    "description": "The type of defect detected",
```
Comment (pj-ms) [inline visionmetrics/key_value_pair/key_value_pair_eval.py:40]: I get the pt that you want to pass the existing schema from MMI api or vision-datasets directly here, but a schema for this evaluation class should be designed from this class's perspective, not restricted by or for the convenience of the downstream application.

Example: if MMI api changes the schema format, do you want to change this library as well with the justification being MMI api updates it?

File: README.md (hunk 16 above)
Comment (juliagongms) [reply README.md]: Great catch, thanks! Updated several other metrics and descriptions as well.

File: visionmetrics/key_value_pair/key_value_pair_eval.py (hunk 15 above)
Comment (juliagongms) [reply visionmetrics/key_value_pair/key_value_pair_eval.py]: Good point. Added more context and information to describe the classes better.

File: visionmetrics/key_value_pair/key_value_pair_eval.py (hunk 17 above)
Comment (juliagongms) [reply visionmetrics/key_value_pair/key_value_pair_eval.py]: Good question. The rationale for using the same schema is not *because* it matches the downstream application's, but moreso that the schema needs to be flexible enough to represent the task of key-value pair extraction with additional returned values like grounding for each key; so, the design is the same because the application's design also took these same factors into account to properly represent the task. Previous versions of the schema would not have enabled the correct logic for inferring metric types. JSON Schema is also a pretty standard format across various applications today.

It is true that the library will likely evolve, but the hope is that the schema is flexible enough to adapt to additional features that we would want to add for this metric.

Comment (pj-ms) [reply visionmetrics/key_value_pair/key_value_pair_eval.py]: "JSON Schema is also a pretty standard format across various applications today."

Not really talking about whether it should be JSON or XML for the schema, but more about what the schema is, for what purpose. For example, is "description" in your schema needed for evaluation? Not saying you should fail the evaluation if "description" exists, but it shouldn't be mentioned/expected from an evaluation class.

Basically what I am asking is to check if everything from the schema in your example makes sense to evaluation. This is an evaluation class at the end of day. 

If the shared schema includes tons of dataset details not relevant to evaluation, a separation would save you from confusing library users and maintenance burden.

Comment (juliagongms) [reply visionmetrics/key_value_pair/key_value_pair_eval.py]: JSON Schema refers not to the fact that it is in JSON format, but specifically this schema definition, which is a standard for API swaggers and also structured data extraction nowadays: https://json-schema.org/understanding-json-schema/reference.

Good point - the "description" key, for instance, is not actually used in evaluation, only provided in the examples. I'll clean up the examples so they only contain the minimal information needed for the evaluation.

Comment (juliagongms) [reply visionmetrics/key_value_pair/key_value_pair_eval.py]: Hi @pj-ms, please take a look at the most recent code, which addresses this issue. Thanks!
